keps/sig-storage/3476-volume-group-snapshot/README.md

```diff
@@ -0,0 +1,1660 @@
+        // Required
+        // VolumeGroupSnapshotRef specifies the VolumeGroupSnapshot object
+        // to which this VolumeGroupSnapshotContent object is bound.
+        VolumeGroupSnapshotRef core_v1.ObjectReference
+
+        // Required
+        VolumeGroupSnapshotDeletionPolicy VolumeGroupSnapshotDeletionPolicy
+
+        // Required
+        Driver string
+
+        // +optional
+        VolumeGroupSnapshotClassName *string
+
+        // Required
+        Source VolumeGroupSnapshotContentSource
+}
+
+// OneOf
+type VolumeGroupSnapshotContentSource struct {
+        // Dynamical provisioning of VolumeGroupSnapshot
+        // +optional
+        VolumeGroupHandle *string
+
+        // Pre-provisioned VolumeGroupSnapshot
+        // +optional
+        VolumeGroupSnapshotHandle *string
+}
+
+Type VolumeGroupSnapshotContentStatus struct {
+        // VolumeGroupSnapshotHandle is a unique id returned by the CSI driver
+        // to identify the VolumeGroupSnapshot on the storage system.
+        // If a storage system does not provide such an id, the
+        // CSI driver can choose to return the VolumeGroupSnapshot name.
+        // +optional
+        VolumeGroupSnapshotHandle *string
+
+        // ReadyToUse becomes true when ReadyToUse on all individual snapshots become true
+        // +optional
+        ReadyToUse *bool
+
+        // +optional
+        CreationTime *int64
+
+        // +optional
+        Error *VolumeGroupSnapshotError
+
+         // List of volume group snapshot contents
+        // +optional
+        VolumeSnapshotContentList []VolumeSnapshotContent
+}
+```
+
+#### VolumeSnapshot and VolumeSnapshotContent
+
+For VolumeSnapshot, we cannot request a VolumeSnapshot to be added to be VolumeGroupSnapshot, therefore VolumeGroupSnapshotName is only in the Status but not in the Spec.
+
+```
+type VolumeSnapshotStatus struct{
+	......
+        // +optional
+        VolumeGroupSnapshotName *string
+	......
+}
+
+type VolumeSnapshotContentStatus struct{
+	......
+        // +optional
+        VolumeGroupSnapshotContentName *string
+	......
+}
+```
+
+### Example Yaml Files
+
+#### Create VolumeGroupSnapshot
+
+Create a VolumeGroupSnapshotClass:
+```
+apiVersion: volumegroup.storage.k8s.io/v1alpha1
+kind: VolumeGroupSnapshotClass
+metadata:
+  name: volumeGroupSnapshotClass1
+spec:
+  parameters:
+     …...
+```
+
+A VolumeGroupSnapshot taken from multiple volumes dynamically:
+```
+apiVersion: volumegroup.storage.k8s.io/v1alpha1
+kind: VolumeGroupSnapshot
+metadata:
+  name: my-group-snapshot
+spec:
+  selector:
+    myapp: postgresql
+  volumeGroupSnapshotClassName: volumeGroupSnapshotClass1
+```
+
+A new external VolumeGroupSnapshot controller will handle VolumeGroupSnapshotClass, VolumeGroupSnapshot, and VolumeGroupSnapshotContent resources. We may need to split this into two controllers, one common controller that handles common functions such as binding, and one sidecar controller that calls the CSI driver.
```

Comment:
I think we are starting to have a binary proliferation problem. Can we add this logic into the existing snapshot binaries?

Reply:
Yes, we can.